lightningd/channel_control.c

```diff
@@ -322,3 +326,100 @@ bool channel_tell_funding_locked(struct lightningd *ld,
 		lockin_complete(channel);
 	return true;
 }
+
+/* Check if we are the fundee of this channel, the channel
+ * funding transaction is still not yet seen onchain, and
+ * it has been too long since the channel was first opened.
+ * If so, we should forget the channel. */
+static void
+check_fundee_no_funding_tx(struct lightningd *ld,
+			   struct channel *channel,
+			   u32 block_height)
+{
+	u32 max_no_funding_tx = 2016;
+
+	/* FIXME: we should be getting max_no_funding_tx
+	 * from an lightningd option, which is why we get
+	 * it as an argument. */
+	(void) ld;
+
+	/* BOLT #2:
+	 *
+	 * A non-funding node (fundee):
+	 *   - SHOULD forget the channel if it does not see the
+	 * funding transaction after a reasonable timeout.
+	 */
+
+	/* Only applies if we are fundee. */
+	if (channel->funder == LOCAL)
+		return;
+
+	/* Does not apply if we already saw the funding tx. */
+	/* FIXME: This is only set when the funding tx is confirmed
+	 * deeply.
+	 * We want to suppress this if the funding tx is at all
+	 * seen (e.g. on mempool) and still remains valid (i.e.
+	 * has not been double spent).
+	 * In practice maybe "confirmed deeply" is sufficient
+	 * approximation of this condition. */
+	if (channel->scid)
+		return;
+
+	/* Not even reached previous starting blocknum.
+	 * (e.g. if --rescan option is used) */
+	if (block_height < channel->first_blocknum)
+		return;
+
+	/* Timeout in blocks not yet reached. */
+	if (block_height - channel->first_blocknum < max_no_funding_tx)
+		return;
+
+	/* Okay, forget it! */
+	/* Report it first. */
+	log_unusual(channel->log,
+		    "Forgetting channel: "
+		    "It has been %"PRIu32" blocks without the "
+		    "funding transaction %s getting deeply "
+		    "confirmed. "
+		    "We are fundee and can forget channel without "
+		    "loss of funds.",
+		    block_height - channel->first_blocknum,
+		    type_to_string(tmpctx, struct bitcoin_txid,
+				   &channel->funding_txid));
+	/* Forget it later, to prevent loops from crashing.
+	 * Allocate timer from channel so if multiple deletions
+	 * by other checks that get added in the future, only
+	 * first timer triggers. */
+	notleak(new_reltimer(&ld->timers, channel, time_from_sec(0),
+			     &delete_channel, channel));
```

Comment:
Hanging off a timer like this is always a hack.   What happens without this?

If it needs to be outside loop, I would change this to a simple "is_fundee_timed_out", then do:

```
const tal_t *expired = tal(NULL, char);

list_for_each(peer) {
    list_for_each(channel) {
        if (is_fundee_timed_out(channel, blockheight))
            tal_steal(expired, channel);
    }
}

/* Now we free them all at once */
tal_free(expired);
```

Reply:
Ha, yes, that is indeed better.

Reply:
Hmm, maybe not safe?  `delete_channel` does more than just `tal_free`, it also deletes the channel from the DB.  However a side effect of `delete_channel` is to also do a `tal_free` of the channel.

How about putting it in an array then having a separate loop to destroy the array?

    struct channel **expired = tal_arr(NULL, struct channel *, 0);

    list_for_each(peer) {
        list_for_each(channel) {
            if (is_fundee_should_forget(ld, channel, block_height)) {
                tal_resize(&expired, tal_count(expired) + 1);
                expired[tal_count(expired) - 1] = channel;
            }
        }
    }
    for (size_t i = 0; i < tal_count(expired); ++i)
        delete_channel(expired[i]);
    tal_free(expired);